be/src/storage/index/inverted/builtin/builtin_inverted_index_iterator.cpp

```diff
@@ -0,0 +1,208 @@
+    auto first_wildcard_pos = search_query->to_string().find('%');
+    if (first_wildcard_pos == std::string::npos) {
+        return Status::InternalError("invalid wildcard query for builtin inverted index");
+    }
+
+    rowid_t lower_rowid = 0;
+    rowid_t upper_rowid = _bitmap_itr->has_null_bitmap() ? _bitmap_itr->bitmap_nums() - 1 : _bitmap_itr->bitmap_nums();
+    std::pair<rowid_t, std::string> lower = std::make_pair(lower_rowid, Slice::min_value().to_string());
+    std::pair<rowid_t, std::string> upper = std::make_pair(upper_rowid, Slice::max_value().to_string());
+
+    if (first_wildcard_pos != 0) {
+        Slice prefix_s(search_query->data, first_wildcard_pos);
+        std::string next_prefix = get_next_prefix(prefix_s);
+
+        auto seek = [&](const Slice& bound) -> StatusOr<std::pair<rowid_t, std::string>> {
+            bool exact_match;
+            auto st = _bitmap_itr->seek_dictionary(&bound, &exact_match);
+            auto cur_ordinal = _bitmap_itr->current_ordinal();
+            if (st.is_not_found()) {
+                // hit the end of dictionary set
+                return std::make_pair(upper_rowid, Slice::max_value().to_string());
+            } else if (!st.ok()) {
+                return st;
+            } else {
+                auto column = ChunkHelper::column_from_field_type(TYPE_VARCHAR, false);
+                size_t read_num = 1;
+                RETURN_IF_ERROR(_bitmap_itr->next_batch_dictionary(&read_num, column.get()));
+                Slice s = down_cast<BinaryColumn*>(column.get())->get_data()[0];
+                return std::make_pair(cur_ordinal, s.to_string());
+            }
+        };
+
+        ASSIGN_OR_RETURN(lower, seek(prefix_s));
+        if (!next_prefix.empty()) {
+            Slice next_prefix_s(next_prefix);
+            ASSIGN_OR_RETURN(upper, seek(next_prefix_s));
+        }
+
+        if (lower.first >= upper.first) {
+            // prefix not found
+            return Status::OK();
+        }
+
+        if (first_wildcard_pos == search_query->size - 1) {
+            // pure prefix query
+            Buffer<rowid_t> hit_rowids(upper.first - lower.first);
+            std::iota(hit_rowids.begin(), hit_rowids.end(), lower.first);
+            RETURN_IF_ERROR(_bitmap_itr->read_union_bitmap(hit_rowids, bit_map));
+            return Status::OK();
+        }
+    }
+
+    RETURN_IF_ERROR(_init_like_context(*search_query));
+    auto predicate = [this](const Column& value_column) -> StatusOr<ColumnPtr> {
+        Columns cols;
+        cols.push_back(&value_column);
+        cols.push_back(this->_like_context->get_constant_column(1));
+        return LikePredicate::like(this->_like_context.get(), cols);
+    };
+    Slice from_value = Slice(lower.second);
+    size_t search_size = upper.first - lower.first;
+    auto res = _bitmap_itr->seek_dictionary_by_predicate(predicate, from_value, search_size);
```

Comment:
If it’s tokenized at write time, is the like predicate on dictionary right?

Reply:
Yes, it performs like on dictionary